feat: add filter to check invalid chars in user input

kenjis opened the pull request:
**Description**
- add a filter to prevent attacks with malformed character encodings and control characters (null byte)

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

kenjis commented on system/Filters/InvalidChars.php:
```diff
@@ -0,0 +1,120 @@
+     */
+    public function before(RequestInterface $request, $arguments = null)
+    {
+        if ($request->isCLI()) {
+            return;
+        }
```
Just because `CSRF` has the code:
https://github.com/codeigniter4/CodeIgniter4/blob/4995720062ede1ec1fd2e5e595dee7abb6e7556e/system/Filters/CSRF.php#L46-L48

kenjis commented on system/Filters/InvalidChars.php:
```diff
@@ -0,0 +1,120 @@
+
+        throw new RuntimeException(
+            'Invalid UTF-8 characters in ' . $this->source . ': ' . $value
+        );
```
I added SecurityException static named constructors.

MGatner reviewed (approved):
Thanks for those changes, looks great!

kenjis commented:
@MGatner I rebased this PR and added 3 commits.
If you're okay, I will merge this.

> Have you benchmarked this or checked Debug Toolbar timelines to see impact?

I saw Debug Toolbar timelines, but small input does not impact.
Before Filters take 0.0x ms.

paulbalandan commented on system/Filters/InvalidChars.php:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @var string
+     */
+    protected $controlCodeRegex = '/\A[\r\n\t[:^cntrl:]]*\z/u';
```
I think this should not be changeable by child classes and should be private?

kenjis replied:
A user may want to change acceptable control codes.
It is by application specification.
So I changed it as protected property.

paulbalandan replied:
👍

paulbalandan commented on user_guide_src/source/incoming/filters.rst:
```diff
@@ -193,6 +193,6 @@ In this example, the array ``['dual', 'noreturn']`` will be passed in ``$argumen
 ****************
 
-Three filters are bundled with CodeIgniter4: ``Honeypot``, ``CSRF``, and ``DebugToolbar``.
+These filters are bundled with CodeIgniter4: ``Honeypot``, ``CSRF``, ``DebugToolbar`` and ``InvalidChars``.
```
```suggestion
The filters bundled with CodeIgniter4 are: ``Honeypot``, ``CSRF``, ``DebugToolbar``, and ``InvalidChars``.
```

paulbalandan commented on tests/system/Filters/InvalidCharsTest.php:
```diff
@@ -0,0 +1,167 @@
+
+    public function stringWithLineBreakAndTabProvider()
+    {
+        return [
```
```suggestion
        yield from [
```

paulbalandan commented on tests/system/Filters/InvalidCharsTest.php:
```diff
@@ -0,0 +1,167 @@
+
+    public function stringWithControlCharsProvider()
+    {
+        return [
```
```suggestion
        yield from [
```

paulbalandan commented on user_guide_src/source/incoming/filters.rst:
```diff
@@ -193,6 +193,6 @@ In this example, the array ``['dual', 'noreturn']`` will be passed in ``$argumen
-Three filters are bundled with CodeIgniter4: ``Honeypot``, ``CSRF``, and ``DebugToolbar``.
+These filters are bundled with CodeIgniter4: ``Honeypot``, ``CSRF``, ``DebugToolbar`` and ``InvalidChars``.
 
 .. note:: The filters are executed in the declared order  that is defined in the config file, but there is one exception to this and it concerns the ``DebugToolbar``, which is always executed last. This is because ``DebugToolbar`` should be able to register everything that happens in other filters.
```
Although not part of the PR but still..
```suggestion
.. note:: The filters are executed in the order defined in the config file. However, if enabled, ``DebugToolbar`` is always executed last because it should be able to capture everything that happens in the other filters.
```

paulbalandan commented on tests/system/Filters/InvalidCharsTest.php:
```diff
@@ -0,0 +1,167 @@
+        ];
+    }
+
+    /**
+     * @dataProvider stringWithControlCharsProvider
+     *
+     * @param string $input
+     */
+    public function testCheckControlStringWithControlCharsCausesException($input)
```
```suggestion
    /**
     * @dataProvider stringWithControlCharsProvider
     */
    public function testCheckControlStringWithControlCharsCausesException(string $input)
```

paulbalandan commented on tests/system/Filters/InvalidCharsTest.php:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * @doesNotPerformAssertions
+     *
+     * @dataProvider stringWithLineBreakAndTabProvider
+     *
+     * @param string $input
+     */
+    public function testCheckControlStringWithLineBreakAndTabReturnsTheString($input)
```
```suggestion
     * @dataProvider stringWithLineBreakAndTabProvider
     */
    public function testCheckControlStringWithLineBreakAndTabReturnsTheString(string $input)
```

MGatner commented:
💪😊👍
